config/azure.go

```diff
@@ -2,13 +2,109 @@ package config
 
 import (
 	"context"
+	"encoding/json"
 	"fmt"
+	"net/url"
+	"strings"
 
+	"github.com/databricks/databricks-sdk-go/apierr"
 	"github.com/databricks/databricks-sdk-go/httpclient"
 	"github.com/databricks/databricks-sdk-go/logger"
 	"golang.org/x/oauth2"
 )
 
+type tokenError struct {
+	message string
+	err     *httpclient.HttpError
+}
+
+func (e *tokenError) Error() string {
+	return e.message
+}
+
+func (e *tokenError) Unwrap() error {
+	sdkErr, ok := apierr.ByStatusCode(e.err.StatusCode)
+	if ok {
+		// this is how we distinguish between bad requests and permission denies
+		return sdkErr
+	}
+	return e.err
+}
+
+func (c *Config) mapAzureError(defaultErr *httpclient.HttpError) error {
+	env := c.Environment()
+	switch defaultErr.Request.Host {
+	case c.hostOrEmpty(env.AzureActiveDirectoryEndpoint()):
+		return c.mapAzureActiveDirectoryError(defaultErr)
+	case c.hostOrEmpty(env.AzureResourceManagerEndpoint()):
+		return c.mapAzureResourceManagerError(defaultErr)
+	default:
+		// Azure MSI endpoint returns not so typed error bodies: `404 page not found`
+		return &tokenError{
+			message: defaultErr.Message,
+			err:     defaultErr,
+		}
+	}
+}
+
+func (c *Config) hostOrEmpty(endpoint string) string {
+	parsedURL, err := url.Parse(endpoint)
+	if err != nil {
+		return ""
+	}
+	return parsedURL.Host
+}
+
+type azureActiveDirectoryErrorResponse struct {
```

Comment:
Can you link to docs for reference where this structure and the corresponding one for ARM responses are defined?

Reply:
i didn't get it from docs - I've made a couple of failing requests through integration tests, looked at debug output and made the mapping. i don't even know if/where these are documented